pkg/agent/handler/accessrequest.go

```diff
@@ -0,0 +1,188 @@
+package handler
+
+import (
+	"encoding/json"
+	"fmt"
+
+	agentcache "github.com/Axway/agent-sdk/pkg/agent/cache"
+	v1 "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/api/v1"
+	mv1 "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/management/v1alpha1"
+	defs "github.com/Axway/agent-sdk/pkg/apic/definitions"
+	prov "github.com/Axway/agent-sdk/pkg/apic/provisioning"
+	"github.com/Axway/agent-sdk/pkg/util"
+	"github.com/Axway/agent-sdk/pkg/watchmanager/proto"
+)
+
+const (
+	provision     = "Provision"
+	deprovision   = "Deprovision"
+	statusErr     = "Error"
+	statusSuccess = "Success"
+	statusPending = "Pending"
+)
+
+type arProvisioner interface {
+	AccessRequestProvision(accessRequest prov.AccessRequest) (status prov.RequestStatus)
+	AccessRequestDeprovision(accessRequest prov.AccessRequest) (status prov.RequestStatus)
+}
+
+type accessRequestHandler struct {
+	prov   arProvisioner
+	cache  agentcache.Manager
+	client client
+}
+
+// NewAccessRequestHandler creates a Handler for Access Requests
+func NewAccessRequestHandler(prov arProvisioner, cache agentcache.Manager, client client) Handler {
+	return &accessRequestHandler{
+		prov:   prov,
+		cache:  cache,
+		client: client,
+	}
+}
+
+func (h *accessRequestHandler) Handle(action proto.Event_Type, _ *proto.EventMeta, resource *v1.ResourceInstance) error {
+	if resource.Kind != mv1.AccessRequestGVK().Kind {
+		return nil
+	}
+
+	ar := &mv1.AccessRequest{}
+	err := ar.FromInstance(resource)
+	if err != nil {
+		return err
+	}
+
+	if ar.Status.Level == statusErr || ar.Status.Level == statusSuccess {
+		return nil
+	}
+
+	app, err := h.getManagedApp(ar)
+	if err != nil {
+		return err
+	}
+
+	req, err := h.newReq(ar, util.GetAgentDetails(app))
+	if err != nil {
+		return err
+	}
+
+	if ar.Status == nil || ar.Status.Level == "" {
+		return fmt.Errorf("unable to provision AccessRequest %s. Status not found", ar.Name)
+	}
+
+	if action == proto.Event_DELETED {
+		h.prov.AccessRequestDeprovision(req)
+		return nil
+	}
+
+	var status prov.RequestStatus
+
+	if ar.Status.Level == statusPending && ar.State.Name == provision {
+		status = h.prov.AccessRequestProvision(req)
+	}
+
+	if ar.Status.Level == statusPending && ar.State.Name == deprovision {
+		status = h.prov.AccessRequestDeprovision(req)
+	}
+
+	s := prov.NewStatusReason(status)
+	ar.Status = &s
+
+	details := util.MergeMapStringInterface(util.GetAgentDetails(ar), status.GetProperties())
+	util.SetAgentDetails(ar, details)
+
+	bts, err := json.Marshal(ar)
+	if err != nil {
+		return err
+	}
+
+	_, err = h.client.UpdateResource(ar.Metadata.SelfLink, bts)
```

Comment:
Since we are not making any change to the spec just the subresources(x-agent-detail and status) are getting updated, we probably do not need the update call

Reply:
The Status field needs to be updated too right?